### ISSUE TITLE
ARM64 builds

### ISSUE THREAD

**Issue by pavel-odintsov** [wkhtmltopdf/packaging#1]
Hello!

As discussed in https://github.com/wkhtmltopdf/wkhtmltopdf/issues/3886 I moved conversation here. 

Regarding distro support I have experience only with Ubuntu LTS and Debian. Both of them work pretty fine on ARM64 arch.

**Comment by ashkulz** [comment]
Do you have the exact sequence of commands I can run to have a fully working OS running under qemu? Would prefer vagrant images, as I need to implement support for that anyway for OS X and Windows builds.

**Comment by ashkulz** [comment]
The question now becomes, which distros and architectures to support? I'll wait for PRs from people who actually need the builds, as it's quite simple to add support ... don't want to create a lot of targets which no one will eventually use, which is something that I guess happens for the x86 builds.

**Comment by ashkulz** [comment]
Closing this issue, as I'll accept PRs for the distros/architectures which are of interest.

**Comment by romaricdrigon** [comment]
Hello,

As AWS announced as AWS just announced [ARM-based EC2 instances)(https://aws.amazon.com/fr/blogs/aws/new-ec2-instances-a1-powered-by-arm-based-aws-graviton-processors/), a lot of people (including myself), having a compatible build would be great.
I'm not quite sure on how to get started with the build process, it would require adding an entry on `build.yml` and running it on an ARM machine?

**Comment by BlindingDark** [comment]
I tried to build on arm64, but some compilation errors have occurred.
```
error: cast from 'JSC::JSCell*' to 'int32_t {aka int}' loses precision [-fpermissive]
```

It seems to be qt4 own bug. http://lists.openembedded.org/pipermail/openembedded-core/2014-November/098726.html


I even tried to change `int32_t` to `int64_t`, solved this problem but caused another problem. It can't pass COMPILE_ASSERT in file `src/3rdparty/webkit/Source/JavaScriptCore/runtime/JSGlobalData.cpp:126`

```cpp
// Enough storage to fit a JSArray, JSByteArray, JSString, or JSFunction.
// COMPILE_ASSERTS below check that this is true.
char storage[64];
COMPILE_ASSERT(sizeof(JSArray) <= sizeof(storage), sizeof_JSArray_must_be_less_than_storage);
```

So I change `char storage[64];` to `char storage[80];`, and compiled successfully.

But when I tried to run it, a segmentation error occurred. I tried to debug it with gdb, seems to be about `JSValue` or something else...

😢 I don't know how to solve it. I'm just a ruby programmer.

---

packaging version is `0.12.5-1`
wkhtmltopdf version is `4.8.7`
build command I am using is

```shell
./build package-docker arm64 ./path/of/wkhtmltopdf
```
Here is the `build.yml` file I used

```yml
# prefix used in determining full image name in "docker-images" and "compile-docker"
docker-prefix: wkhtmltopdf/packaging:0.12_

# command to run for enabling foreign-arch support in currently running kernel (>= 4.8)
qemu-cmd:      docker run --rm --privileged fkrull/qemu-user-static enable

# all targets supported by the docker-related commands, keys are as follows:
#
#   source:         Dockerfile used to build image
#   output:         one of deb, rpm, tar or 7z
#   args:           build-time docker args (optional)
#   arch:           used as arch specified to fpm (for deb or rpm output)
#   depend:         dependent packages specified to fpm (for deb or rpm output)
#   qt_config:      override the qt configuration used for building Qt (optional)
#   cross_compile:  use given prefix for cross-compilation (optional)
#   qemu:           set to true if qemu emulation is required (optional)
#
docker-targets:
  arm64:
    source: docker/Dockerfile.debian
    args:
      from: arm64v8/ubuntu:16.04
    output: 7z
    qemu:   true
    arch:   arm64
    depend: >
      ca-certificates
      fontconfig
      libc6
      libfreetype6
      libjpeg62-turbo
      libpng16-16
      libssl1.1
      libstdc++6
      libx11-6
      libxcb1
      libxext6
      libxrender1
      xfonts-75dpi
      xfonts-base
      zlib1g

# Qt configure arguments
qt-config:
  common: >
    -opensource
    -confirm-license
    -fast
    -release
    -static
    -graphicssystem raster
    -webkit
    -exceptions
    -xmlpatterns
    -system-zlib
    -system-libpng
    -system-libjpeg
    -no-libmng
    -no-libtiff
    -no-accessibility
    -no-stl
    -no-qt3support
    -no-phonon
    -no-phonon-backend
    -no-opengl
    -no-declarative
    -no-script
    -no-scripttools
    -no-sql-ibase
    -no-sql-mysql
    -no-sql-odbc
    -no-sql-psql
    -no-sql-sqlite
    -no-sql-sqlite2
    -no-mmx
    -no-3dnow
    -no-sse
    -no-sse2
    -no-multimedia
    -nomake demos
    -nomake docs
    -nomake examples
    -nomake tools
    -nomake tests
    -nomake translations
  debug:
    -debug
    -webkit-debug
  docker: >
    -silent
    -xrender
    -largefile
    -iconv
    -openssl-linked
    -no-javascript-jit
    -no-rpath
    -no-dbus
    -no-nis
    -no-cups
    -no-pch
    -no-gtkstyle
    -no-nas-sound
    -no-sm
    -no-xshape
    -no-xinerama
    -no-xcursor
    -no-xfixes
    -no-xrandr
    -no-mitshm
    -no-xinput
    -no-xkb
    -no-glib
    -no-gstreamer
    -no-icu
    -no-openvg
    -no-xsync
    -no-audio-backend
    -no-sse3
    -no-ssse3
    -no-sse4.1
    -no-sse4.2
    -no-avx
    -no-neon
  mingw-w64: >
    -silent
    -openssl-linked
    -no-reduce-exports
    -no-rpath
    -no-pch
    -xplatform win32-g++-4.6
  macos-carbon: >
    -silent
    -reduce-exports
    -no-framework
    -no-dwarf2
    -openssl-linked
    -largefile
    -no-rpath
    -carbon
    -platform unsupported/macx-clang
  macos-cocoa: >
    -silent
    -reduce-exports
    -no-framework
    -no-dwarf2
    -openssl-linked
    -largefile
    -no-rpath
    -cocoa
    -platform unsupported/macx-clang-libc++
  msvc: >
    -mp
    -qt-style-windows
    -no-style-cleanlooks
    -no-style-windowsxp
    -no-style-windowsvista
    -no-style-plastique
    -no-style-motif
    -no-style-cde
    -openssl-linked

# docker image which provides fpm
fpm-image:      wkhtmltopdf/fpm:20190502

# parameters for building linux packages via fpm (deb/rpm)
fpm-params:
  name:         wkhtmltox
  description:  convert HTML to PDF and various image formats using QtWebkit
  license:      LGPLv3
  vendor:       wkhtmltopdf
  maintainer:   Ashish Kulkarni <ashish@kulkarni.dev>
  url:          https://wkhtmltopdf.org/
  prefix:       /usr/local
  category:     utils
```

**Comment by ashkulz** [comment]
This has been addressed via #66

**Comment by ashkulz** [comment]
Please see the [latest release](https://github.com/wkhtmltopdf/packaging/releases) which has packages for a lot of ARM64 targets. Please add a comment if you face any problems in using the packages :+1:

**Comment by ashkulz** [comment]
A release candidate for the **0.12.6** release is [now available for download](https://wkhtmltopdf.org/downloads.html#release-candidate), which should contain changes which possibly address this issue.

Would appreciate downloading the package and reporting back if any issues are encountered during testing. Assuming all goes well, I plan to release 0.12.6 on the 2-year anniversary of the previous release i.e. June 11, 2020.

**Comment by DanteLogan** [comment]
Is this topic still oen for discussion as I am trying to install odoo14 on my raspberry pi and i am guessing i would need to find all the arm64 package equivalents

**Comment by ashkulz** [comment]
Not sure what the issue is, there are "pure"  arm64 as well as raspbian builds available for 0.12.6 -- debian 10 builds should work with 11.
